### PR TITLE
Improvements to tackle wrong expression outputs

### PR DESCRIPTION
**Pull request by poke1024**
see https://github.com/mathics/Mathics/issues/602

### CODE REVIEW

**Comment by poke1024** [comment]
You can get write fancy with an `$OutputSizeLimit` of `1000` now:

example1

Eventually, the limit will still kick in:

example2

We could issue a warning via `evaluate.message` in these cases, telling the user to bump the `$OutputSizeLimit`. We could also use `MakeBoxes` to make the omit tags like `<<2>>` appear in a slightly color/background.

**Comment by poke1024** [comment]
With warning:

example3

File: mathics/core/expression.py [hunk 1]
```diff
@@ -1155,6 +1167,21 @@ def block(tex, only_subsup=False):
...
+    def output_cost(self):
+        name = self.get_head_name()
+
+        if name in ('System`ImageBox', 'System`GraphicsBox', 'System`Graphics3DBox'):
+            return 0  # always display
+
+        cost = sum(leaf.output_cost() for leaf in self.leaves)
+
+        if name == 'System`List':
+            return 2 + cost + len(self.leaves)  # {a, b, c}
```
Comment (sn6uv) [inline mathics/core/expression.py]: would it be better to compute the sum cost of leaves? Also, what about the size of ", " between leaves?

File: mathics/core/expression.py [hunk 2]
```diff
@@ -1155,6 +1167,21 @@ def block(tex, only_subsup=False):
...
+        elif name in _layout_boxes:
+            return cost
+        else:
+            return cost + 2 + self.head.output_cost() + len(self.leaves)  # XYZ[a, b, c]
```
Comment (sn6uv) [inline mathics/core/expression.py]: here also

Comment (poke1024) [inline mathics/core/expression.py]: `cost` is the sum cost of leaves (I renamed it now). `len(self.leaves)` measures the size of ",", which could be one or two (it assumes for now). we could pass the separator size into `make_boxes` and then pass it on to `output_cost`, but it's bulky and then should we really measure it that way? If `$OutputSizeLimit` is understood as number of printable characters (without spaces and layout elements like fractions, parentheses, ...), then it makes sense to always measure 1 for ",".

Comment (sn6uv) [inline mathics/core/expression.py]: In that case I guess technically it should be `max(len(self.leaves) - 1, 0)`.

**Comment by sn6uv** [comment]
It seems like MMA kernels don't even observe `$OutputSizeLimit` only the notebook interface does and the full boxes are still generated anyway.

**Comment by Piruzzolo** [comment]
So our approach is more efficient! Good to know

File: mathics/core/expression.py [hunk 3]
```diff
@@ -1173,15 +1176,20 @@ def output_cost(self):
...
+        if name in _layout_boxes:
             return cost_of_leaves
         else:
-            return cost_of_leaves + 2 + self.head.output_cost() + separator_cost * len(self.leaves)  # XYZ[a, b, c]
+            separator_cost = 1  # i.e. ","
+            n_separators = max(len(leaves) - 1, 0)
+            total_cost = 2 + cost_of_leaves + separator_cost * n_separators  # {a, b, c}, [a, b, c]
+
+            if name != 'System`List':
+                total_cost += self.head.output_cost() + separator_cost * n_separators
```
Comment (sn6uv) [inline mathics/core/expression.py]: doesn't this count the separators twice now?

**Comment by sn6uv** [comment]
This will need to be rebased on a revert of 8e14668.

**Comment by mmatera** [comment]
I made a rebase in #1371. @poke1024 if you have the time and the will, look at it.

**Pull request closed without merging**